perception/multi_object_tracker/src/processor/input_manager.cpp

```diff
@@ -0,0 +1,343 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "multi_object_tracker/processor/input_manager.hpp"
+
+#include <cassert>
+
+namespace multi_object_tracker
+{
+///////////////////////////
+/////// InputStream ///////
+///////////////////////////
+InputStream::InputStream(rclcpp::Node & node, uint & index) : node_(node), index_(index)
+{
+}
+
+void InputStream::init(const InputChannel & input_channel)
+{
+  // Initialize parameters
+  input_topic_ = input_channel.input_topic;
+  long_name_ = input_channel.long_name;
+  short_name_ = input_channel.short_name;
+  is_spawn_enabled_ = input_channel.is_spawn_enabled;
+
+  // Initialize queue
+  objects_que_.clear();
+
+  // Initialize latency statistics
+  latency_mean_ = 0.2;  // [s] (initial value)
+  latency_var_ = 0.0;
+  interval_mean_ = 0.0;  // [s] (initial value)
+  interval_var_ = 0.0;
+
+  latest_measurement_time_ = node_.now();
+  latest_message_time_ = node_.now();
+}
+
+bool InputStream::getTimestamps(
+  rclcpp::Time & latest_measurement_time, rclcpp::Time & latest_message_time) const
+{
+  if (!isTimeInitialized()) {
+    return false;
+  }
+  latest_measurement_time = latest_measurement_time_;
+  latest_message_time = latest_message_time_;
+  return true;
+}
+
+void InputStream::onMessage(
+  const autoware_auto_perception_msgs::msg::DetectedObjects::ConstSharedPtr msg)
+{
+  const DetectedObjects objects = *msg;
+  objects_que_.push_back(objects);
+  if (objects_que_.size() > que_size_) {
+    objects_que_.pop_front();
+  }
+
+  // update the timing statistics
+  rclcpp::Time now = node_.now();
+  rclcpp::Time objects_time(objects.header.stamp);
+  updateTimingStatus(now, objects_time);
+
+  // trigger the function if it is set
+  if (func_trigger_) {
+    func_trigger_(index_);
+  }
+}
+
+void InputStream::updateTimingStatus(const rclcpp::Time & now, const rclcpp::Time & objects_time)
+{
+  // Update latency statistics
+  // skip initial messages for the latency statistics
+  if (initial_count_ > 4) {
```

Comment:
I prefer adding some explanation to these magic numbers. Like: `constexpr int INITIAL_COUNT_THRESHOLD_LATENCY = 4;`.

Reply:
Thank you for your comment.
the constants were defined and explained.

https://github.com/autowarefoundation/autoware.universe/pull/6820/commits/b62399378f67d07e949cd6174a33146616a53bb2